### PR TITLE
HOSTEDCP-2220: Autonode karpenter

### PR DESCRIPTION
**Pull request by enxebre**
**What this PR does / why we need it**:
This introduces support for an autoNode feature via karpenter:
- It introduces a self contained karpenter-operator that is built within the HO image, so its lifecycle is fully decouple from the CPO/OCP version.
- Let the HO reconcile the karpenter-operator within the controlplane namespace
- The karpenter operator manages karpenter deployment management side and manages CRDs, default nodeclass, and clobber "service owned" fields guest side.
- Introduces a --auto-node flag for the CLI which also causes a new arn generation.
- Expose autoNode with karpenter behind a feature gate within the HC API

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

File: karpenter-operator/controllers/karpenter/karpenter_controller.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+	upsert.CreateOrUpdateProvider
+}
+
+func (r *Reconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, managementCluster cluster.Cluster) error {
+	r.ManagementClient = managementCluster.GetClient()
+	r.GuestClient = mgr.GetClient()
+	r.CreateOrUpdateProvider = upsert.New(false)
+
+	// First install the CRDs so we can create a watch below.
+	mgr.GetClient()
```
Comment (muraee) [inline karpenter-operator/controllers/karpenter/karpenter_controller.go:56]: the return value is ignored, is this needed?

File: karpenter-operator/controllers/karpenter/karpenter_controller.go [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+
+	if err := r.reconcileEC2NodeClassOwnedFields(ctx, userDataSecret); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.reconcileEC2NodeClassDefault(ctx, userDataSecret, hcp); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.reconcileKarpenter(ctx, hcp); err != nil {
```
Comment (muraee) [inline karpenter-operator/controllers/karpenter/karpenter_controller.go:161]: I can't find the definition of this function anywhere :)

File: karpenter-operator/controllers/karpenter/karpenter_controller.go (hunk 1 above)
Comment (enxebre) [reply karpenter-operator/controllers/karpenter/karpenter_controller.go]: no needed,  I dropped it. Thanks!

**Comment by enxebre** [comment]
/override "Red Hat Konflux / hypershift-operator-main-on-pull-request"

**Pull request merged**